### ISSUE TITLE
AsyncCompletions.create not recognized as coroutine by inspect.iscoroutinefunction

### ISSUE THREAD

**Issue by liucheng-ms** [openai/openai-python#996]
### Confirm this is an issue with the Python library and not an underlying OpenAI API

- [X] This is an issue with the Python library

### Describe the bug

The `AsyncCompletions.create` method in the `openai-python` library is an asynchronous method that one would expect to be recognized as a coroutine by the `inspect.iscoroutinefunction`. However, checks against this method currently return `False`, suggesting that it is not being identified as an async function. This could lead to issues in asynchronous programming where the correct identification of coroutines is crucial for event loops and async frameworks to schedule and execute tasks appropriately.

### To Reproduce

1. Install the latest version of the OpenAI Python library.
2. Use the `inspect.iscoroutinefunction` to check if `AsyncCompletions.create` is recognized as a coroutine.
3. The result of this check is `False` instead of the expected `True`.

### Code snippets

```Python
import inspect
import openai

is_coroutine = inspect.iscoroutinefunction(openai.resources.AsyncCompletions.create)
print(f'Is `openai.resources.AsyncCompletions.create` a coroutine function? {is_coroutine}')
```

### OS

Windows 11

### Python version

Python v3.9.18

### Library version

openai v1.5.0

**Comment by RobertCraigie** [comment]
Realistically this is not a reliable way to determine if a function is async or not, the only completely reliable way to do so is to call it and check if the return value is awaitable

```py
value = client.completions.create(...)
print(inspect.isawaitable(value))
```

There does appear to be a way for us to mark that the method is a coroutine but this is only available in Python 3.12+ which makes it unusable for us unfortunately.

> This could lead to issues in asynchronous programming where the correct identification of coroutines is crucial for event loops and async frameworks to schedule and execute tasks appropriately.

Can you share any concrete examples?

**Comment by liucheng-ms** [comment]
Thank you for your prompt response.

I appreciate your explanation regarding the limitations of the `inspect.iscoroutinefunction` method and the potential future solution with `inspect.markcoroutinefunction` in Python 3.12. However, I've identified that the issue stems from the use of a synchronous decorator on the `AsyncCompletions.create` async function within the OpenAI Python library. This decorator is not preserving the async nature of the function it decorates, leading to it not being recognized as a coroutine by `asyncio.iscoroutinefunction`.

This is particularly problematic for frameworks like FastAPI, which rely on the `iscoroutinefunction` check to determine if a function should be awaited or called directly. If the library's async function is not recognized as a coroutine, FastAPI will attempt to call it directly rather than awaiting it, which could result in a runtime error or unexpected behavior.
https://github.com/tiangolo/fastapi/blob/e7756ae7dcffa0b73767d354ca158fd6b7bc87ed/fastapi/dependencies/utils.py#L492

To maintain compatibility with asynchronous frameworks and avoid potential issues, it's crucial that decorators used on async functions are also asynchronous. Here's a suggestion for an adjustment to the `required_args` decorator to ensure it correctly handles async functions:
https://github.com/openai/openai-python/blob/3ad4e8bc9d89d7a81586bf598289ff62b0a339b9/src/openai/_utils/_utils.py#L203

```py
def required_args(*variants: Sequence[str]) -> Callable[[CallableT], CallableT]:
    # ... (existing code for decorator setup)

    async def async_wrapper(*args: object, **kwargs: object) -> object:
        # ... (move the logic that was previously in the wrapper here)
        return await func(*args, **kwargs)

    def inner(func: CallableT) -> CallableT:
        # ... (existing code for inner function setup)

        if asyncio.iscoroutinefunction(func):
            @functools.wraps(func)
            async def wrapper(*args: object, **kwargs: object) -> object:
                # ... (call the async_wrapper instead of previous logic)
                return await async_wrapper(*args, **kwargs)
        else:
            @functools.wraps(func)
            def wrapper(*args: object, **kwargs: object) -> object:
                # ... (existing synchronous wrapper logic)
                return func(*args, **kwargs)

        return wrapper  # type: ignore

    return inner
```

By introducing an asynchronous wrapper, we can ensure that the decorated function retains its coroutine properties, allowing async frameworks to manage it appropriately.

Could we consider implementing a fix like this to preserve the async nature of functions after they are decorated? This would help maintain compatibility with async frameworks and prevent potential issues for users.

Thank you for considering this request.

**Comment by antont** [comment]
Just a +1 to at least investigate this, and thanks a lot @liucheng-ms for explaining the issue.

We're also using FastAPI for our endpoints that use the OpenAI lib then, but we are not bit by this as there's always some own functions which are registered as the FastAPI(/Starlette) routes and then those call OpenAI deeper in the code. And our own functions are ofc declared async so they are detected correctly.

**Comment by rattrayalex** [comment]
@liucheng-ms can you share an example of the code you're trying to write?

**Comment by liucheng-ms** [comment]
@rattrayalex Thank you for your continued engagement on this matter. As per your request, I have included a snippet of the code I am working on that illustrates the issue at hand:

```py
def inject_operation_headers(f):  
    @functools.wraps(f)  
    def wrapper(*args, **kwargs):  
        # Inject headers from customized header, overwrite injected header with headers from kwargs.  
        injected_headers = get_customized_headers()  
        original_headers = kwargs.get("headers") if IS_LEGACY_OPENAI else kwargs.get("extra_headers")  
        if original_headers and isinstance(original_headers, dict):  
            injected_headers.update(original_headers)  
        kwargs.update(headers=injected_headers) if IS_LEGACY_OPENAI else kwargs.update(extra_headers=injected_headers)  
        return f(*args, **kwargs)  
  
    @functools.wraps(f)  
    async def async_wrapper(*args, **kwargs):  
        # Inject headers from customized header, overwrite injected header with headers from kwargs.  
        injected_headers = get_customized_headers()  
        original_headers = kwargs.get("headers") if IS_LEGACY_OPENAI else kwargs.get("extra_headers")  
        if original_headers and isinstance(original_headers, dict):  
            injected_headers.update(original_headers)  
        kwargs.update(headers=injected_headers) if IS_LEGACY_OPENAI else kwargs.update(extra_headers=injected_headers)  
        return await f(*args, **kwargs)  
  
    if inspect.iscoroutinefunction(f):
        return async_wrapper  
    else:  
        return wrapper

openai.resources.chat.AsyncCompletion = inject_operation_header(openai.resources.chat.AsyncCompletion)
```
The intended behavior of the `inject_operation_headers` decorator is to enhance the function it decorates with additional headers, regardless of whether the function is synchronous or asynchronous. The issue arises when we use `inspect.iscoroutinefunction` to check if `openai.resources.chat.AsyncCompletion.create` is a coroutine function after decoration. The check should return `True` for asynchronous functions, allowing the `async_wrapper` to be used. However, we are currently encountering the opposite—`False` is returned, and the synchronous `wrapper` is applied instead.

Even though the function seems to work in its current state, this could be due to the pre-processing nature of the header's injection logic. If we were to include post-processing logic in the wrapper, we might not see the same success. The importance of this distinction extends beyond my use case to anyone using the library in conjunction with asynchronous frameworks or code that relies on proper coroutine identification.

Since OpenAI's Python library is widely used and forms a part of critical infrastructure for many applications, it is essential that it robustly supports both synchronous and asynchronous paradigms. In scenarios like mine, developers need confidence that async functions will continue to be treated as such, even after being passed through decorators.

Thank you for considering this request and for any effort you can allocate to address this in the library. I am looking forward to a resolution that supports the async nature of functions reliably.

**Comment by rattrayalex** [comment]
Ah. 

For that purpose, I recommend using [httpx's event handlers](https://www.python-httpx.org/advanced/#event-hooks) to manipulate headers for all requests, e.g.,

```py
def add_custom_headers(request):
    injected_headers = get_customized_headers()
    for k, v in injected_headers.items():
        request.headers[k] = v

http_client = httpx.Client(event_hooks={'request': [add_custom_headers]})

client = AsyncOpenAI(http_client=http_client)
```

Alternatively, if your needs are simple enough, you could even use the `default_headers` kwarg in the client:

```py
client = AsyncOpenAI(default_headers=get_customized_headers())
```

I agree it would be nice for these functions to be introspectable as coroutines, but realistically I don't think it's something we're going to be able to prioritize anytime soon (possibly not until 3.11 hits EOL in late 2027). 

Until then, if you need introspection without calling the function, you should be able to rely on the fact that all methods on our resources that extend `AsyncAPIResource` are async.

Related issues:

- https://github.com/python/cpython/issues/98590
- https://github.com/python/cpython/issues/100224

Given that any async function which is wrapped by a decorator cannot be detected as async with `inspect.iscoroutinefunction`, and wrapping functions with decorators is common in Python (and, not being considered a breaking change, could happen at any time to any function in any library), I would not recommend using this function to introspect third-party libraries like this one.

Accordingly, I'll go ahead and close this for now, but again we may add a marker for it in the future.

**Comment by liucheng-ms** [comment]
@rattrayalex, thank you for your detailed response. I appreciate the alternative solutions you've provided, but I believe the core issue here is not about finding a workaround for my specific use case. Instead, it's about addressing a potential pitfall in the library that could impact other developers.

While I understand that the library has other priorities, I would like to clarify a few points:

1. **Common Practice**: While it's common to decorate functions in Python, it's not common or advisable to use a synchronous decorator on an asynchronous function. This practice can lead to unexpected behavior, as we've seen with the `inspect.iscoroutinefunction` check.

2. **Python Fundamental Issue**: The issues you've linked from the Python repository are not entirely relevant to this discussion. The first one is about `functools.partial`, which is not being used here, and it has been fixed in 3.11. The second one is about an async decorator  on sync function being identified as async, which is a coder's problem, not a Python problem. In our case, we're using a sync decorator on an async function, which leads to `inspect.iscoroutinefunction` returning False. These are generally considered a code-level issue rather than a language limitation.

3. **Decorator Adjustment**: As I've suggested in my previous comment, adjusting the `required_args` decorator to handle async functions correctly could be a potential solution. This adjustment ensures that the decorated function retains its coroutine properties, allowing async frameworks to manage it appropriately.

Given the library's widespread use and its potential role in critical infrastructure, it would be beneficial for the library to uphold the async standards and conventions of the Python ecosystem.

I understand that there may be constraints on what can be prioritized, but I urge the maintainers to reconsider the impact of this issue. Perhaps it could be addressed incrementally or when resources allow. In the meantime, could we document this behavior prominently, so developers are aware of the current limitation and can plan accordingly?

Thank you for your time and consideration. I am looking forward to a resolution that enhances the library's compatibility with the async ecosystem.

**Comment by umarbutler** [comment]
I managed to workaround this by inspecting the source code of functions to check whether they contain a line that begins with `async def` followed by the function’s name (and also do not have a line beginning with `def` and the function’s name), as illustrated below:
```python
def is_async(func: Callable) -> bool:
    """Determine whether a callable is asynchronous."""
    
    # If `inspect.iscoroutinefunction` identifies the callable as asynchronous, then return `True`. If it doesn't, then try to search for a line that begins (ignoring preceeding whitespace) with `async def` followed by the callable's name and a `(` character inside its source code, returning `True` if such a line is found and there is no such line beginning with `def` (indicating that asynchronous and synchronous functions with the same name were defined in the same block as the callable).
    if inspect.iscoroutinefunction(func):
        return True
    
    try:
        source = inspect.getsource(func)
        
    except (OSError, TypeError):
        return False

    has_async_def = re.search(r'^\s*async\s+def\s+' + re.escape(func.__name__) + r'\s*\(', source, re.MULTILINE)
    
    if has_async_def:    
        has_sync_def = re.search(r'^\s*def\s+' + re.escape(func.__name__) + r'\s*\(', source, re.MULTILINE)
        
        if not has_sync_def:
            return True

    return False
```

The only catch is that this is not absolutely guaranteed to capture all asynchronous functions. `inspect.getsource` does not work for builtins, for example. Also, if you define two functions within a single decorator that have the same name, then this will also not work. Nor will it work if you change the `__name__` attribute of a function later on. 

Nevertheless, this method has sufficed for my uses and may be of some benefit to others.
